gax/operation.h

```diff
@@ -0,0 +1,138 @@
+
+namespace google {
+namespace gax {
+
+/**
+ * Operation is a user-visible primitive used to support Long Running Operations
+ * (LROs). LROs are service methods that may take a 'long time' (anywhere from
+ * seconds to days), and instead of blocking for that long return an Operation
+ * object or an error status. The Operation may contain a result and always
+ * contains metadata.
+ *
+ * A service method that returns `google.longrunning.Operation` and that has
+ * the correct annotations for its result and metadata types will cause the
+ * generator to define a client method that returns
+ * gax::StatusOr<gax::Operation<ResponseT, MetadataT>>.
+ *
+ * Example usage:
+ *
+ * @code
+ * gax::StatusOr<gax::Operation<Foo, FooMeta>> res =
+ * client.GetBigFoo(getBigFooRequest);
+ * if(!res) {
+ *   ...
+ * }
+ *
+ * operationsClient = client.OperationsClient();
+ * Operation<Foo, FooMeta> op = *std::move(res);
+ * while(!op.Done()) {
+ *   gax::Status stat = operationsClient.Update(op);
+ *   if(isPermanentFailure(stat)) {
+ *     ...
+ *   }
+ *   std::cout << "Foo request " << getBigFooRequest().name() << " is " <<
+ *             op.Metadata().percent() << "% done.";
+ *   std::this_thread::sleep(sleep_period());
+ * }
+ *
+ * gax::StatusOr<Result> terminus = op.Result();
+ * if(!terminus) {
+ *   ...
+ * }
+ * Result r = *std::move(terminus);
+ * @endcode
+ */
+template <
+    typename ResponseT, typename MetadataT,
+    typename std::enable_if<
+        std::is_base_of<protobuf::Message, ResponseT>::value, int>::type = 0,
+    typename std::enable_if<
+        std::is_base_of<protobuf::Message, MetadataT>::value, int>::type = 0>
+class Operation final {
+ public:
+  // Note: the constructor is intended to be used by GAPIC generated code, not
+  // users.
+  Operation(google::longrunning::Operation op) : op_(std::move(op)) {}
+
+  /**
+   * @name Return the service-provided name of the underlying
+   * google.longrunning.Operation
+   */
+  std::string const& Name() const { return op_.name(); }
+
+  /**
+   * @name If Operation::Done(), return the underlying Response, or an error
+   * code if a problem occurred. Otherwise return an error indicating that the
+   * Operation has not completed.
+   */
+  gax::StatusOr<ResponseT> Result() const {
+    if (!Done()) {
+      return gax::Status{gax::StatusCode::kUnknown,
+                         "operation has not completed=" + Name()};
+    } else if (op_.has_error()) {
+      return gax::Status{static_cast<gax::StatusCode>(op_.error().code()),
+                         op_.error().message()};
+    } else {
+      auto const& any = op_.response();
+      if (!any.Is<ResponseT>()) {
+        return gax::Status{gax::StatusCode::kUnknown,
+                           "invalid result in operation=" + Name()};
+      }
+
+      ResponseT result;
+      any.UnpackTo(&result);
+      return std::move(result);
+    }
+  }
+
+  /**
+   * @name Return the most recent metadata received from the service.
```

Comment:
`@name` is for [naming groups](http://www.doxygen.nl/manual/commands.html#cmdname), depending how you configured Doxygen the brief description of a function is either `@brief` or simply the first line in the Doxygen comment.

Reply:
Good to know! I've never written for Doxygen before, so all the annotations were cargo culted.